src/goto-instrument/contracts/dynamic-frames/dfcc_contract_functions.cpp

```diff
@@ -66,6 +69,29 @@ dfcc_contract_functionst::dfcc_contract_functionst(
 
   spec_functions.to_spec_frees_function(
     spec_frees_function_id, nof_frees_targets);
+
+  // instrument the generated functions for side-effect checking
+  std::set<irep_idt> function_pointer_contracts;
+
+  instrument.instrument_function(
+    spec_assigns_function_id,
+    dfcc_loop_contract_modet::NONE,
+    function_pointer_contracts);
+
+  INVARIANT(
+    function_pointer_contracts.empty(),
+    id2string(spec_assigns_function_id) +
+      " shall not contain calls to " CPROVER_PREFIX "obeys_contract");
+
+  instrument.instrument_function(
+    spec_frees_function_id,
+    dfcc_loop_contract_modet::NONE,
+    function_pointer_contracts);
+
+  INVARIANT(
+    function_pointer_contracts.empty(),
+    id2string(spec_frees_function_id) +
+      " shall not contain calls to " CPROVER_PREFIX "obeys_contract");
```

Comment:
I think we should have a variant of `instrument_function` that takes only two arguments and includes that `INVARIANT`.

Reply:
factored